SearchControl: Deprecate onClose prop

Vrishabhsk opened the pull request:
## What?
- Add `@deprecated` for `onClose` function prop for `SearchControl component`
- Update `README.md`, `JSDoc` and `StoryBook` with `_Note: this prop is deprecated._`
- When `onClose` prop is provided, execute `deprecated` from `@wordpress/deprecated`
- Update `test` to expect `console.warn` when `onClose` is provided + fix `should should` typo

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
- https://github.com/WordPress/gutenberg/issues/65334
> The onClose prop is not used anymore in the Gutenberg app
> The prop itself is causing confusion to consumers (https://github.com/WordPress/gutenberg/issues/64921),

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- `Soft Deprecates` the `onClose` prop
- Add `deprecation` notice in the `SearchControl's docs`
- Execute `deprecated` when `onClose` is provided
- Add `test` to expect `deprecation warning`

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Clone the branch into the `plugins` directory of your `WordPress` installation
```
git clone https://github.com/Vrishabhsk/gutenberg/tree/update/deprecate-onClose
```
- Build the files
```
cd gutenberg && npm install & npm build
```
- Activate the `Gutenberg`plugin via the `Plugins Admin Page`.
- Create a custom block
```
npx @wordpress/create-block@latest demo --variant=dynamic
```
- Modify the `edit.js` to use `SelectControl` component
```
export default function Edit() {
	return (
		<p {...useBlockProps()}>
			<h1>Shut up</h1>
			<InspectorControls>
				<PanelBody title="Form Settings" initialOpen={false}>
					<PanelRow>
						<SearchControl
							label={__("Search posts")}
							onChange={(val) => {}}
						/>
					</PanelRow>
				</PanelBody>
			</InspectorControls>
		</p>
	);
}
```
- To test the deprecated log, use the following snippet : [`Deprecated Hook`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-deprecated/#hook)

Vrishabhsk committed:
Soft deprecate onClose function prop and execute deprecated

Vrishabhsk committed:
Update README and StoryBook JSDoc

Vrishabhsk committed:
Expect console.warn due to deprecated func call and fix typo

mirka commented on packages/components/src/search-control/index.tsx:
```diff
@@ -31,6 +32,14 @@ function SuffixItem( {
 	}
 
+	if ( onClose ) {
+		deprecated( 'onClose function prop', {
+			since: '6.7',
+			version: '',
+			plugin: 'wp.components.SearchControl',
+		} );
```
- Let's phrase it this way so the component name is visible in the console warning. (Apparently the `plugin` string won't show unless [`version` is truthy](https://github.com/WordPress/gutenberg/blob/e86152bc8b45d5fe213ed688f145e82f73bb2fd4/packages/deprecated/src/index.js#L46-L48)).
- This isn't going to land in 6.7, so it'll be the next release.
- We can just omit the `version` property.

```suggestion
		deprecated( '`onClose` prop in wp.components.SearchControl', {
			since: '6.8',
		} );
```

mirka commented on packages/components/src/search-control/stories/index.story.tsx:
```diff
@@ -52,6 +52,8 @@ Default.args = {
 };
 
 /**
+ * _Note: this prop is deprecated._
+ *
  * When an `onClose` callback is provided, the search control will render a close button
  * that will trigger the given callback.
  *
```
I think we can remove this entire story, since we don't want to recommend it.

mirka commented on packages/components/src/search-control/test/index.tsx:
```diff
@@ -72,13 +72,14 @@ describe( 'SearchControl', () => {
 			expect( onChangeSpy ).toHaveBeenLastCalledWith( '' );
 		} );
 
-		it( 'should should render a Close button (instead of Reset) when onClose function is provided', async () => {
```
Thanks for the cleanup! 🧹

mirka commented on packages/components/src/search-control/types.ts:
```diff
@@ -40,6 +40,8 @@ export type SearchControlProps = Pick< InputControlProps, 'help' | 'value' > & {
 	 *
 	 * Use this if you want the button to trigger your own logic to close the search field entirely,
 	 * rather than just clearing the input value.
+	 *
+	 * @deprecated
```
Let's also add a `@ignore` tag, which will suppress this from the Storybook props table.

```suggestion
	 *
	 * @deprecated
	 * @ignore
```

mirka reviewed (commented):
Looking good, thank you for contributing!

In case it's useful when contributing to wp-components in the future, we usually just spin up the Storybook (`npm run storybook:dev`) when testing things like this, without going through the hassle of making a test plugin 🙂

mirka commented:
Could you also add a [changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entry please? Thanks!

Vrishabhsk committed:
Address PR feedback

Remove WithOnClose storybook variation. Refactor deprecation message. Add @ignore for onClose in types.ts. Add deprecations section in CHANGELOG.md

ciampo commented on packages/components/src/search-control/test/index.tsx:
```diff
@@ -72,13 +72,14 @@ describe( 'SearchControl', () => {
 				<Component onChange={ onChangeSpy } onClose={ onCloseSpy } />
 			);
 
+			expect( console ).toHaveWarned();
```
Nit: can we use `toHaveWarnedWith` and specify the deprecation message? It would make the test easier to understand and would guard against any other unwanted warnings

Vrishabhsk committed:
Replace toHaveWarned with toHaveWarnedWith

mirka reviewed (approved):
Perfect, thank you!

mirka commented:
## ✍️ Dev Note

### SearchControl: Deprecated `onClose` prop

This prop was originally intended for adding a custom click handler to the suffix button to close the search field entirely, rather than just clear the input value.

The pattern of repurposing the search clear button as a search close button is no longer used in WordPress, and is no longer recommended as a UI pattern since it can be confusing to users.

If you were relying on this prop, we recommend adding a separate close button to your UI.
